// elysia-prometheus@1.0.0/dist/index.mjs
// src/index.ts
import { Elysia } from "elysia";
import {
  collectDefaultMetrics,
  Registry,
  Counter,
  Histogram
} from "prom-client";
var DEFAULT_OPTIONS = {
  metricsPath: "/metrics",
  durationBuckets: [3e-3, 0.03, 0.1, 0.3, 1.5, 10],
  staticLabels: {},
  dynamicLabels: {},
  useRoutePath: true
};
var index_default = (userOptions = {}) => {
  const opts = { ...DEFAULT_OPTIONS, ...userOptions };
  const register = new Registry();
  collectDefaultMetrics({ register });
  const reservedLabels = /* @__PURE__ */ new Set(["method", "path", "status"]);
  const allLabels = { ...opts.staticLabels, ...opts.dynamicLabels };
  for (const label of Object.keys(allLabels)) {
    if (reservedLabels.has(label)) {
      throw new Error(`Label '${label}' is reserved`);
    }
  }
  const labelNames = [
    "method",
    "path",
    "status",
    ...Object.keys(opts.staticLabels),
    ...Object.keys(opts.dynamicLabels)
  ];
  const httpRequestCounter = new Counter({
    name: "http_requests_total",
    help: "Total HTTP requests count",
    labelNames,
    registers: [register]
  });
  const httpRequestDuration = new Histogram({
    name: "http_request_duration_seconds",
    help: "HTTP request duration in seconds",
    labelNames,
    buckets: opts.durationBuckets,
    registers: [register]
  });
  const getStatusCode = (ctx) => {
    if (typeof ctx.response === "object" && ctx.response !== null && "code" in ctx.response && typeof ctx.response.code === "number") {
      return ctx.response.code.toString();
    }
    if (ctx.set.status) {
      return ctx.set.status.toString() ?? "unknown";
    }
    return "500";
  };
  function getLabels(ctx) {
    const path = opts.useRoutePath ? ctx.route || ctx.path : ctx.path;
    const labels = {
      method: ctx.request.method,
      path: normalizePath(path),
      status: getStatusCode(ctx),
      ...opts.staticLabels
    };
    for (const [key, fn] of Object.entries(opts.dynamicLabels)) {
      labels[key] = fn(ctx);
    }
    return labels;
  }
  function normalizePath(path) {
    return path.replace(/\/\d+([\/?]|$)/g, "/:id$1");
  }
  return new Elysia({ name: "prometheus" }).derive({ as: "global" }, (ctx) => ({
    endTimer: httpRequestDuration.startTimer(getLabels(ctx))
  })).onAfterHandle({ as: "global" }, (ctx) => {
    if (ctx.path.endsWith(opts.metricsPath)) return;
    httpRequestCounter.inc(getLabels(ctx));
    ctx.endTimer(getLabels(ctx));
  }).onError({ as: "global" }, (ctx) => {
    if (!ctx.endTimer) return;
    if (ctx.path.endsWith(opts.metricsPath)) return;
    httpRequestCounter.inc(getLabels(ctx));
    ctx.endTimer(getLabels(ctx));
  }).get(opts.metricsPath, async () => {
    return new Response(await register.metrics(), {
      headers: { "Content-Type": register.contentType }
    });
  });
};
export {
  index_default as default
};
